go-client/TestSetupAbstraction.go

```diff
@@ -0,0 +1,210 @@
+		err := os.Mkdir(serverDir, 0700)
+		if err != nil {
+			panic(fmt.Sprintf("failed to create directory for the test-setup-abstraction-server. Cause: %v", err.Error()))
+		}
+	}
+	serverFile := path.Join(serverDir, serverJar)
+	if _, err := os.Stat(serverFile); os.IsNotExist(err) {
+		out, err := os.Create(serverFile)
+		defer func() {
+			err = out.Close()
+			if err != nil {
+				panic(fmt.Sprintf("failed to close server file. Cause: %v", err.Error()))
+			}
+		}()
+		if err != nil {
+			panic(fmt.Sprintf("failed to create file for downloading test-setup-abstraction-server. Cause: %v", err.Error()))
+		}
+		resp, err := http.Get("https://github.com/exasol/exasol-test-setup-abstraction-server/releases/download/" + serverVersion + "/" + serverJar)
+		if err != nil {
+			panic(fmt.Sprintf("failed to download test-setup-abstraction-server. Cause: %v", err.Error()))
+		}
+		defer resp.Body.Close()
+		_, err = io.Copy(out, resp.Body)
+		if err != nil {
+			panic(fmt.Sprintf("failed to download test-setup-abstraction-server. Cause: %v", err.Error()))
+		}
+	}
+	return serverFile
+}
+
+func getServerPort(output *bytes.Buffer, errorStream *bytes.Buffer) int {
+	for counter := 0; counter < 500; counter++ {
+		pattern, err := regexp.Compile("Server running on port: (\\d+)\n")
+		if err != nil {
+			panic(err)
+		}
+		result := pattern.FindSubmatch(output.Bytes())
+
+		if len(result) != 0 {
+			portString := string(result[1])
+			port, err := strconv.ParseInt(portString, 10, 32)
+			if err != nil {
+				panic(err)
+			}
+			return int(port)
+		}
+		time.Sleep(1 * time.Second)
+	}
+	fmt.Println(errorStream.String())
+	fmt.Println(output.String())
+	panic("failed to start server. The server did not print a port number.")
+}
+
+func waitForServer(serverProcess *exec.Cmd, errorStream *bytes.Buffer, stopped *bool,
+	stoppedMutex *sync.Mutex) {
+	err := serverProcess.Run()
+	if err != nil && !isStopped(stopped, stoppedMutex) { //after we killed the thread we expect an error
+		fmt.Println(errorStream.String())
+		panic(fmt.Sprintf("failed to start test-setup-abstraction server. Cause: %v", err.Error()))
+	}
+}
+
+func isStopped(stopped *bool, stoppedMutex *sync.Mutex) bool {
+	stoppedMutex.Lock()
+	defer stoppedMutex.Unlock()
+	return *stopped
+}
+
+func (testSetup *TestSetupAbstraction) Stop() {
+	testSetup.stoppedMutex.Lock()
+	*testSetup.stopped = true
+	testSetup.stoppedMutex.Unlock()
+	err := testSetup.server.Process.Signal(os.Kill)
+	if err != nil {
+		panic(fmt.Sprintf("failed to stop test-setup-abstraction server. Cause: %v", err.Error()))
+	}
+}
+
+func (testSetup *TestSetupAbstraction) getConnectionInfo() *ConnectionInfo {
+	var connectionDetails ConnectionInfo
+	testSetup.makeApiRequest("GET", "connectionInfo", &connectionDetails, nil)
+	return &connectionDetails
+}
+
+func (testSetup *TestSetupAbstraction) makeApiRequest(method string, address string, jsonResult interface{}, payload url.Values) {
+	client := http.Client{}
```

Comment:
I think we could. But I'm not sure if it's an omprovement. Since it's not a constructor we probalby just allocate an empty struct here. So I guess this should be cheap.

Reply:
👍